ql/src/java/org/apache/hadoop/hive/ql/io/avro/AvroGenericRecordReader.java

```diff
@@ -88,22 +88,39 @@ public AvroGenericRecordReader(JobConf job, FileSplit split, Reporter reporter)
       gdr.setExpected(latest);
     }
 
-    if (split.getLength() == 0) {
-      this.isEmptyInput = true;
-      this.start = 0;
-      this.reader = null;
+    org.apache.avro.file.FileReader<GenericRecord> nonFinalReader = null;
+    try {
+      if (split.getLength() == 0) {
+        this.isEmptyInput = true;
+        this.start = 0;
+        this.reader = null;
+      }
+      else {
+        this.isEmptyInput = false;
+        nonFinalReader = new DataFileReader<GenericRecord>(new FsInput(split.getPath(), job), gdr);
+        this.reader = nonFinalReader;
+        this.reader.sync(split.getStart());
+        this.start = reader.tell();
+      }
+      this.stop = split.getStart() + split.getLength();
+      this.recordReaderID = new UID();
+
+      this.writerTimezone = extractWriterTimezoneFromMetadata(job, split, gdr);
+      this.writerProleptic = extractWriterProlepticFromMetadata(job, split, gdr);
     }
-    else {
-      this.isEmptyInput = false;
-      this.reader = new DataFileReader<GenericRecord>(new FsInput(split.getPath(), job), gdr);
-      this.reader.sync(split.getStart());
-      this.start = reader.tell();
+    catch (Exception e) {
+      if (nonFinalReader != null) {
+        try {
+          nonFinalReader.close();
+        }
+        catch (Exception closeException) {
+          if (closeException != e) {
+            e.addSuppressed(closeException);
+          }
+        }
+      }
+      throw e;
```

Comment:
Why not close the reader in finally?

Reply:
We cannot. On positive flow (where no exeception is thrown) the readers should remain open and be assigned to `AvroGenericRecordReader.reader` instance field.

Reply:
Got it. Thanks!